File: backend/poketeambuilder/src/main/java/albert/lozano/poketeambuilder/domain/Trainer.java
/**
 * @author Albert Lozano Blasco
 * @version 1.0
 */

package albert.lozano.poketeambuilder.domain;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.*;

import java.util.Date;
import java.util.List;

/**
 * Trainer class. Defines a trainer.
 */
@Entity
@Table(name = "Trainer")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Trainer {
    // Attributes
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long trainerId;

    @NotNull
    @Column
    private String username;

    @NotNull
    @Column
    private String password;

    @NotNull
    @Column
    private String name;

    @NotNull
    @Column
    private String firstSurname;

    @Column
    private String secondSurname;

    @NotNull
    @Column
    private String email;

    @NotNull
    @Column
    private String phone;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @Column
    private Date createdDate;

    @Column
    private boolean theme;

    @OneToMany(mappedBy = "trainer", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Team> teams;

    @OneToMany(mappedBy = "trainer", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Comment> comments;

    @Column
    private String bio;

    public boolean getTheme() {
        return this.theme;
    }
}

File: frontend/poketeambuilder/lib/data/models/trainer.dart
import 'dart:convert';

import 'package:poketeambuilder/data/models/team.dart';

class Trainer {
  final String username;
  String password;
  String name;
  String firstSurname;
  String secondSurname;
  final String email;
  String phone;
  final DateTime createdDate;
  bool theme;
  String bio;

  Trainer({
    required this.name,
    required this.firstSurname,
    required this.secondSurname,
    required this.email,
    required this.phone,
    required this.username,
    required this.password,
    required this.createdDate,
    required this.theme,
    required this.bio,
  });

  Trainer.withUsername(this.username)
      : password = '',
        name = '',
        firstSurname = '',
        secondSurname = '',
        email = '',
        phone = '',
        createdDate = DateTime.now(),
        theme = false,
        bio = '';

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'firstSurname': firstSurname,
      'secondSurname': secondSurname,
      'email': email,
      'phone': phone,
      'username': username,
      'password': password,
      'createdDate': createdDate.toIso8601String(),
      'theme': theme,
      'bio': bio,
    };
  }

  factory Trainer.fromJson(String json) {
    final decodedJson = jsonDecode(json) as Map<String, dynamic>;

    return Trainer(
      name: decodedJson['name'] as String,
      firstSurname: decodedJson['firstSurname'] as String,
      secondSurname: decodedJson['secondSurname'] as String,
      email: decodedJson['email'] as String,
      phone: decodedJson['phone'] as String,
      username: decodedJson['username'] as String,
      password: decodedJson['password'] as String,
      createdDate: DateTime.parse(decodedJson['createdDate'] as String),
      theme: decodedJson['theme'] as bool,
      bio: decodedJson['bio'] as String,
    );
  }

}
